fetch/request.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fetch
+
+import (
+	"net/http"
+)
+
+func clone(src []string) []string {
+	if src == nil {
+		return nil
+	}
+	dst := make([]string, len(src))
+	copy(dst, src)
+	return dst
+}
+
+// RequestTweaker mutates an http.Request to set the refer(r)er, add custom
+// HTTP headers, and so on.
+type RequestTweaker interface {
+	// Tweak mutates req. parent is the request which spawned req. If req is
+	// a request to fetch a subresource, parent is typically the one used to
+	// fetch the main resource. parent can be nil (e.g. when req is a request
+	// to fetch a main resource) and should not be mutated.
+	Tweak(req *http.Request, parent *http.Request) error
+}
+
+// DefaultRequestTweaker is a RequestTweaker used by default.
+var DefaultRequestTweaker RequestTweaker = SetReferer()
+
+// RequestTweakerSequence consists of a series of RequestTweakers.
+type RequestTweakerSequence []RequestTweaker
+
+// Tweak invokes all tweakers in order. It fails immediately when some tweaker
+// returns an error, in which case the subsequent tweakers will not run.
+func (seq RequestTweakerSequence) Tweak(req, parent *http.Request) error {
+	for _, tweaker := range seq {
+		if err := tweaker.Tweak(req, parent); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// SetReferer sets the Referer HTTP header with the parent request URL.
+func SetReferer() RequestTweaker {
+	return &setReferer{}
+}
+
+type setReferer struct{}
+
+func (*setReferer) Tweak(req, parent *http.Request) error {
```

Comment:
I believe a value receiver is preferred here, per https://github.com/golang/go/wiki/CodeReviewComments#receiver-type.
We may need to modify constructors as well.

Reply:
I don't think so; I'd apply "when in doubt, use a pointer receiver." If I read the linked doc correctly, there are only a clj;.e  cases where we should use value receivers:

* The receiver is a map, func, or chan
* The receiver is not a map, func, or chan.
* The method doesn't mutate the receiver.
* The receiver doesn't contain anything synchronous.
* The receiver is just a trivial struct, not large.
* There are no changes made to the original receiver.
* The receiver is effectively immutable.
* The receiver isn't analogous to int or str.

Reply:
I checked golang's source code and found plenty of both. I'd defer to you, or probably we need to ask some expert, exactly who should have done that kind of reviewing in the first place?

Reply:
If we could see both patterns, I'd go with the current code for now ― I've used pointer receivers for all empty structs in this codebase. I'll consider reaching out to some Go expert(s), but AFAICT there's no official channel to receive what you mean by "that kind of reviewing."

Reply:
SGTM